worker/task.go

```diff
@@ -0,0 +1,284 @@
+)
+
+type TaskRun struct {
+	TaskId          string                       `json:"taskId"`
+	RunId           uint                         `json:"runId"`
+	SignedDeleteUrl string                       `json:"-"`
+	TaskClaim       queue.TaskClaimResponse      `json:"-"`
+	TaskReclaim     queue.TaskReclaimResponse    `json:"-"`
+	Definition      queue.TaskDefinitionResponse `json:"-"`
+	QueueClient     queueClient
+	plugin          plugins.TaskPlugin
+	log             *logrus.Entry
+	payload         interface{}
+	sync.RWMutex
+	context        *runtime.TaskContext
+	controller     *runtime.TaskContextController
+	sandboxBuilder engines.SandboxBuilder
+	sandbox        engines.Sandbox
+	resultSet      engines.ResultSet
+}
+
+// Abort will set the status of the task to aborted and abort the task execution
+// environment if one has been created.
+func (t *TaskRun) Abort() {
+	t.Lock()
+	defer t.Unlock()
+	if t.context != nil {
+		t.context.Abort()
+	}
+	if t.sandbox != nil {
+		t.sandbox.Abort()
+	}
+}
+
+// Cancel will set the status of the task to cancelled and abort the task execution
+// environment if one has been created.
+func (t *TaskRun) Cancel() {
+	t.log.Info("Cancelling task")
+	t.Lock()
+	defer t.Unlock()
+	if t.context != nil {
+		t.context.Cancel()
+	}
+	if t.sandbox != nil {
+		t.sandbox.Abort()
+	}
+}
+
+// ExceptionStage will report a task run as an exception with an appropriate reason.
+// Tasks that have been cancelled will not be reported as an exception as the run
+// has already been resolved.
+func (t *TaskRun) ExceptionStage(status runtime.TaskStatus, err error) {
+	if t.context.IsCancelled() {
+		return
+	}
+
+	var reason string
+	switch err.(type) {
+	case engines.MalformedPayloadError:
+		reason = "malformed-payload"
+	case engines.InternalError:
+		reason = "internal-error"
+	default:
+		reason = "worker-shutdown"
+	}
+
+	update := TaskStatusUpdate{
+		Task:          t,
+		Status:        status,
+		Reason:        reason,
+		WorkerId:      t.TaskClaim.WorkerId,
+		ProvisionerId: t.TaskClaim.Status.ProvisionerId,
+		WorkerGroup:   t.TaskClaim.WorkerGroup,
+	}
+
+	<-UpdateTaskStatus(update, t.QueueClient, t.log)
+	return
+}
+
+// Run is the entrypoint to executing a task run.  The task payload will be parsed,
+// plugins created, and the task will run through each of the stages of the task
+// life cycle.
+//
+// Tasks that do not complete successfully will be reported as an exception during
+// the ExceptionStage.
+func (t *TaskRun) Run(pluginManager plugins.Plugin, engine engines.Engine, context *runtime.TaskContext, contextController *runtime.TaskContextController) {
+	t.Lock()
+	t.context = context
+	t.controller = contextController
+	t.Unlock()
+
+	defer t.DisposeStage()
+
+	jsonPayload := map[string]json.RawMessage{}
+	if err := json.Unmarshal(t.Definition.Payload, &jsonPayload); err != nil {
+		t.context.LogError(fmt.Sprintf("Invalid task payload. %s", err))
+		t.ExceptionStage(runtime.Errored, err)
+		return
```

Comment:
Okay, how about `ExceptionStage` also returns the error if it is an internal error... And then we return the error from `Run`.

Specifically, I think we'll want to shutdown (kill the ec2 instance) in the face of internal errors, or have some logic that counts the number of internal errors and shutdown after it goes really bad...

Regardless we probably don't want to deal with that here so let's propagate internal errors up (after reporting the task as an exception and cleaning up, etc...)...

That way this also becomes `return t.ExceptionStage(runtime.Errored, err)` which is nice :)